src/lib.rs

```diff
@@ -0,0 +1,1012 @@
+			return Err(Error::NotRunning);
+		}
+
+		let runtime = run_lock.as_ref().unwrap();
+
+		// Stop wallet sync
+		runtime.stop_wallet_sync.store(true, Ordering::Release);
+
+		// Stop networking
+		runtime.stop_networking.store(true, Ordering::Release);
+		self.peer_manager.disconnect_all_peers();
+
+		// Drop the runtime, which stops the background processor and any possibly remaining tokio threads.
+		*run_lock = None;
+		Ok(())
+	}
+
+	fn setup_runtime(&self) -> Result<Runtime, Error> {
+		let tokio_runtime =
+			tokio::runtime::Builder::new_multi_thread().enable_all().build().unwrap();
+
+		// Setup wallet sync
+		let chain_access = Arc::clone(&self.chain_access);
+		let sync_cman = Arc::clone(&self.channel_manager);
+		let sync_cmon = Arc::clone(&self.chain_monitor);
+		let sync_logger = Arc::clone(&self.logger);
+		let stop_wallet_sync = Arc::new(AtomicBool::new(false));
+		let stop_sync = Arc::clone(&stop_wallet_sync);
+
+		tokio_runtime.spawn(async move {
+			let mut rounds = 0;
+			loop {
+				if stop_sync.load(Ordering::Acquire) {
+					return;
+				}
+				// As syncing the on-chain wallet is much more time-intesive, we only sync every
+				// fifth round.
+				if rounds == 0 {
+					let now = Instant::now();
+					match chain_access.sync_wallet().await {
+						Ok(()) => log_info!(
+							sync_logger,
+							"On-chain wallet sync finished in {}ms.",
+							now.elapsed().as_millis()
+						),
+						Err(_) => log_error!(sync_logger, "On-chain wallet sync failed"),
+					}
+				}
+				rounds = (rounds + 1) % 5;
+
+				let confirmables = vec![
+					&*sync_cman as &(dyn Confirm + Sync),
+					&*sync_cmon as &(dyn Confirm + Sync),
+				];
+				let now = Instant::now();
+				match chain_access.sync(confirmables).await {
+					Ok(()) => log_info!(
+						sync_logger,
+						"Lightning wallet sync finished in {}ms.",
+						now.elapsed().as_millis()
+					),
+					Err(e) => log_error!(sync_logger, "Lightning wallet sync failed: {}", e),
+				}
+				tokio::time::sleep(Duration::from_secs(5)).await;
+			}
+		});
+
+		// Setup networking
+		let peer_manager_connection_handler = Arc::clone(&self.peer_manager);
+		let listening_port = self.config.listening_port;
+		let stop_networking = Arc::new(AtomicBool::new(false));
+		let stop_listen = Arc::clone(&stop_networking);
+
+		tokio_runtime.spawn(async move {
+			let listener =
+				tokio::net::TcpListener::bind(format!("0.0.0.0:{}", listening_port)).await.expect(
+					"Failed to bind to listen port - is something else already listening on it?",
+				);
+			loop {
+				if stop_listen.load(Ordering::Acquire) {
+					return;
+				}
+				let peer_mgr = Arc::clone(&peer_manager_connection_handler);
+				let tcp_stream = listener.accept().await.unwrap().0;
+				tokio::spawn(async move {
+					lightning_net_tokio::setup_inbound(
+						Arc::clone(&peer_mgr),
+						tcp_stream.into_std().unwrap(),
+					)
+					.await;
+				});
+			}
+		});
+
+		// Regularly reconnect to channel peers.
```

Comment:
Hm, not quite sure what you're referring to? Looking them up to connect to them? But if we have a channel, they should be in the peer store?

Reply:
Mh, but wouldn't that mean that we need to keep track when we added the peer and see if any update since then changed the address? If so, we probably should default to not store addresses for public nodes separately and always try to look them up from gossip?